Add lesson8 (#8)
ReDI-School/js-berlin-2019-fall: slides/lesson8.md

```

---

### Quiz - Question 4

* How can we set the text in an html element?

---

### Quiz - Question 4

* How can we set the text in an html element?

```js
myElement.textContent = "Hello";
```

---

### Quiz - Question 5

* How do we execute JavaScript when a button is clicked?

---

### Quiz - Question 5

* How do we execute JavaScript when a button is clicked?

```html
<button onclick="myFunction()">Click me!</button>
```

---

### Variable lifetime

Variables live only in their scope (the block of code surrounded by curly braces):

```js
function foo();
{
  let x = 42;
  // x is available
}

// x is NOT available
```

---

### Variable lifetime (2)

```js
function foo();
{
  let x = 42;
  // x is available
  if (x >= 42) {
    let y = 7331;
    // y and x are available
  }
  // only x is available, y is NOT available
}

// x and y are NOT available
```

---

### Global scope

Variables declared _outside_ a code block are in the global scope and are globally available:

```js
let buttonClickCount = 0;

function onButtonClick() {
  buttonClickCount += 1;
  console.log("you clicked the button " + buttonClickCount + " times");
}
```

---

### Objects

> * Objects are a collection of *keys* and *values*.

Reviewer: ```suggestion
* Objects are a collection of *key* and *value* pairs.
```

* The key is for us a label (like a variable name). You will also see this called a "property". Keys unlock what is behind the door.
* The value is the type represented by the key. What is behind the door!
* Together they create a pair. Remember: objects are made of Pairs.
They come together. They are the building blocks of objects.

---

### Remember: Pairs

![Pears](images/pears.png) <!-- .element height="600px" width="600px" style="display: block; margin: 0 auto" -->

---

### Example

```js
let person = {
  firstName: 'John',  // key: firstName - value: 'John'
  lastName: 'Doe'     // key: lastName  - value: 'Doe'
};
```

> * Objects start end end with curly braces `{}`

Reviewer: ```suggestion
* Objects start and end with curly braces `{}`
```

* Write some facts about you and print them to console:

```js
let person = {
  firstName: 'Bob',
  age: 32,
  isStudent: true
};

console.log("Hi, I am " + person.firstName + " and ...");
```

---

### Why objects?

Imagine writing a function that prints details about a person without using objects:

```js
function printPerson(firstName, lastName, age, isStudent, favoriteFood, streetAddress, city, country, phone, fax, facebookName) {
  console.log("Hi I am " + firstName + " ... ");
}
```

---

### Why objects? (2)

Now imagine the same function when using objects:

```js
function printPerson(person) {
  console.log("Hi, I am " + person.firstName + " ... ");
}
```

All the person's details are stored in one variable, much less to type!

---

### Objects in Objects

* A 'value' in an object can itself be an object:

```js
let person = {
  name: "John",
  address: {
    street: 'Am Nordbahnhof',
    city: 'Berlin'
  }
};
```

Quiz: How do we access the person's city?

---

### Objects in Objects

```js
let person = {
  name: "John",
  address: {
    street: 'Am Nordbahnhof',
    city: 'Berlin'
  }
};

console.log(person.address.city);
```

---

### Functions in Objects

> So far, we had `number`, `string`, `boolean` and `object` as values. How about `function`?

Reviewer: ```suggestion
So far, we had `number`, `string`, `boolean` and `object` as types of values. How about `function`?
```

let myValue = myInputElement.value;
```

---

### Quiz - Question 4

* How can we set the text in an html element?

---

### Quiz - Question 4

* How can we set the text in an html element?

```js
myElement.textContent = "Hello";
```

---

### Quiz - Question 5

* How do we execute JavaScript when a button is clicked?

---

### Quiz - Question 5

* How do we execute JavaScript when a button is clicked?

```html
<button onclick="myFunction()">Click me!</button>
```

---

### Variable lifetime

Variables live only in their scope (the block of code surrounded by curly braces):

```js
function foo();
{
  let x = 42;
  // x is available
}

// x is NOT available
```

---

### Variable lifetime (2)

```js
function foo();
{
  let x = 42;
  // x is available
  if (x >= 42) {
    let y = 7331;
    // y and x are available
  }
  // only x is available, y is NOT available
}

// x and y are NOT available
```

---

### Global scope

Variables declared _outside_ a code block are in the global scope and are globally available:

```js
let buttonClickCount = 0;

function onButtonClick() {
  buttonClickCount += 1;
  console.log("you clicked the button " + buttonClickCount + " times");
}
```

---

> ### Objects

Reviewer: Increase spacing between sentences

> ### Objects

Reviewer: Increase spacing between sentences

> * Objects are a collection of *key* and *value* pairs.

Reviewer: ```suggestion
* Objects are a collection of *properties*
* A property is a pair of a *key* and a *value*.
```

> * The key is for us a label (like a variable name). You will also see this called a "property". Keys unlock what is behind the door.

Reviewer: ```suggestion
* The key is for us a label (like a variable name). Keys unlock what is behind the door.
```

> * The value is the type represented by the key. What is behind the door!

Reviewer: ```suggestion
* The value is the value represented by the key. What is behind the door!
```

> * Together they create a pair. Remember: objects are made of Pairs.

Reviewer: Remove this sentence. Already mentioned properties above.

> They come together. They are the building blocks of objects.

Reviewer: Remove this sentence too. Mentioned above.

---

### Quiz - Question 4

* How can we set the text in an html element?

```js
myElement.textContent = "Hello";
```

---

### Quiz - Question 5

* How do we execute JavaScript when a button is clicked?

---

### Quiz - Question 5

* How do we execute JavaScript when a button is clicked?

```html
<button onclick="myFunction()">Click me!</button>
```

---

### Variable lifetime

Variables live only in their scope (the block of code surrounded by curly braces):

```js
function foo();
{
  let x = 42;
  // x is available
}

// x is NOT available
```

---

### Variable lifetime (2)

```js
function foo();
{
  let x = 42;
  // x is available
  if (x >= 42) {
    let y = 7331;
    // y and x are available
  }
  // only x is available, y is NOT available
}

// x and y are NOT available
```

---

### Global scope

Variables declared _outside_ a code block are in the global scope and are globally available:

```js
let buttonClickCount = 0;

function onButtonClick() {
  buttonClickCount += 1;
  console.log("you clicked the button " + buttonClickCount + " times");
}
```

---

### Objects

* Objects are a collection of *properties*
* A property is a pair of a *key* and a *value*.

> * The key is for us a label (like a variable name). Keys unlock what is behind the door.

Reviewer: * Every piece of data is stored in the object with a key (like a variable name).
* Keys unlock what is behind the door.
* The type of key is string.

> * The value is the value represented by the key. What is behind the door!

Reviewer: ```suggestion
* The value is the content accessed by the key. What is behind the door!
* The type of value can be any type
```
